### ISSUE TITLE
step scaling

### ISSUE THREAD

**Issue by kofifus** [chafey/cornerstone#114]
we are having an issue where some big images have artifacts when scaled down.

This is a general canvas issue and a common solution is  step scaling
see: https://jsfiddle.net/AbdiasSoftware/M4cTx/

Is there a way to make cornerstone step scale images ?

thanks!

**Comment by swederik** [comment]
Hello,

There is no way to make Cornerstone do this at the moment. It sounds like an interesting technique but it might have a significant effect on performance.

**Comment by kofifus** [comment]
it will mean keeping an original copy of the image and on each scaling creating scaled copy to work with (As in the fiddle), zooming will definitely end up slower (but maybe ok) but we get the improved image quality plus I imagine window leveling will become very fast as the image is actually much smaller.

On some images (ie 6 mega pixels CR with a grid in front of the plate), the different in quality is quite dramatic.

**Comment by jpambrun** [comment]
Yeah. This technique is usually referred as mip mapping. Basically, you need to apply a low pass filter when re-sampling in accordance with the Nyquist sampling theorem in order to avoid aliasing. This is the functional equivalent to interpolation for downscaling.

With that said, implementing may be a nice to have, but reduce performance at startup (to build the pyramid) and would take an additional 1/3 memory.  The patterns that produce aliasing artefacts are not all that common in medical imaging and they should be examined as close to full resolution for diagnostic purposes.

Once the pyramid computed, it could indeed increase the performance of W/L when zoomed out. 

It would be nice if the browser could just apply it by itself just like it does interpolation...

**Comment by jpambrun** [comment]
@kofifus, Do you have a DICOM file that shows these artefacts that you could share by chance?

**Comment by jpambrun** [comment]
Never-mind, I created a sample :
[aliasing512x512.dcm.zip](https://github.com/chafey/cornerstone/files/947592/aliasing512x512.dcm.zip)

Also, a video of the issue:
[alias.mp4.zip](https://github.com/chafey/cornerstone/files/947602/alias.mp4.zip)

**Comment by kofifus** [comment]
ok great, I will try to also get my dicom file to you.

so is there any ideas about solving this with cornerstone ?

thx!

**Comment by jpambrun** [comment]
It would be nice if you can get any real case showing this issue. I'm not
convinced, given the tradeoffs, that this would add significant clinical
value.

**Comment by kofifus** [comment]
Here is a sample image:
     https://www.dropbox.com/s/7gxsqajpovnij0a/Cornerstone_aliasing.zip?dl=0

You will see the artifacts as you zoom in and out.

The image was taken with a grid in front of the plate (I can find out the name of the procedure).

Thanks!

**Comment by jpambrun** [comment]
I could see the artefacts between 30% and 40% magnification.


I suppose mip mapping could be added behind a flag as it increases resource
requirements for a limited return.


Unfortunately, I don't think this is on anybody's road map... Contributions
are welcomed.


JF

**Comment by chafey** [comment]
FWIW, most viewers do not worry about this use case as images must be
displayed at full resolution (or higher) for interpretation.

**Comment by FMCalisto** [comment]
As already mentioned here, I suggest the step scaling approach, which involves keeping an original copy of the image and creating scaled copies to work with. This technique is usually referred to as mip mapping, and would require additional memory and potentially reduce performance at startup, but could improve image quality and potentially increase performance for certain functions such as window leveling when zoomed out. It seems that this feature is not currently on anyone's roadmap for development, but we can try to work around.
